Safe freeing of ids for HA slaves

lutovich opened the pull request.

Commit:
Slave now dont see record reads contaminated by record reuse

MASTER
Master now buffer ids for a specified time, called the safe zone. Within this safe zone window
records are guaranteed not to be reused.

SLAVE
UpdatePuller makes sure no transactions break the safe zone boundary
by killing transactions that fall outside during pull update.

The transactions store the latest_applied_transaction_commit_time when
they start and `TransactionCommitingResonseUnpacker` kills the transactions
as new transactions (pulled from master) get applied by comparing
the stored commit_time with the new latest_applied_transaction_commit_time
after applying the batch.

If the batch itself spans over a greater time than what fits inside the safe zone,
no more transacions can be started on slave until the entire batch is applied.

Safe zone is configurable through ... a setting

SECONDARY CHANGES
Added knowledge about last tx commit timestamp to MetaDataStore and StoreMigrator

Commit:
TimeUtil use proper TimeUnit

Commit:
Turn off short-lived locks

Commit:
Clean up around TransactionTerminatedException

Commit:
Remove volatile terminated flag to only use volatile termination reason

Commit:
Improve blocking of new txs in KernelTransactions

lutovich committed:
Added tests for blocking of new transactions

lutovich committed:
Cleanups around id buffering

 * added tests here and there
 * completed and removed couple TODOs
 * remove unused classes and revert not needed changes

lutovich committed:
Smaller methods in TransactionCommittingResponseUnpacker

lutovich committed:
Extract HA's IdReuseEligibility in a top level class

Commit:
Test correctness of reading property chains with concurrent record reuse

Commit:
Fixup! Small clean up

Commit:
Transfer last transactionCommitTimestamp on upgrade

Commit:
Formatting

chrisvest commented:
> SLAVE
> UpdatePuller makes sure no transactions break the safe zone boundary
> by killing transactions that fall outside during pull update.

Can you explain this part in more detail?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -534,7 +534,7 @@ public void transactionRecovered( long txId )
                     dependencies.satisfyDependency( new LegacyIndexApplierLookup.Direct( legacyIndexProviderLookup ) );
 
-            boolean safeIdBuffering = FeatureToggles.flag( getClass(), "safeIdBuffering", false );
+            boolean safeIdBuffering = FeatureToggles.flag( getClass(), "safeIdBuffering", true );
```
Feature toggles should be stored in `static final` fields to enable constant propagation optimisations, and also to avoid parsing strings and allocating objects all the time.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -851,7 +851,7 @@ public StoreReadLayer storeLayer()
     private Factory<StoreStatement> storeStatementFactory( final NeoStores neoStores )
     {
-        final LockService lockService = FeatureToggles.flag( getClass(), "propertyReadLocks", true ) ?
+        final LockService lockService = FeatureToggles.flag( getClass(), "propertyReadLocks", false ) ?
```
Again, `static final` that feature toggle. For great justice!

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/KernelTransaction.java:
```diff
@@ -112,16 +113,16 @@
+     * @return {@link Status} if {@link #markForTermination(Status)} has been invoked, otherwise {@code null}.
      */
-    boolean shouldBeTerminated();
+    Status terminationReason();
```
`getReasonIfTerminated` would be a clearer name, I think.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/Status.java:
```diff
@@ -121,6 +121,10 @@ public Code code()
                 " and so this transaction was rolled back although it may have looked like it was going to be " +
                 "committed" ),
+        Outdated( TransientError, "Transaction has seen state which has been invalidated by applied updates while " +
+                "transaction was active. Transaction should succeed if retried." ),
```
s/should/may/

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/Status.java:
```diff
@@ -121,6 +121,10 @@ public Code code()
                 "committed" ),
+        Outdated( TransientError, "Transaction has seen state which has been invalidated by applied updates while " +
+                "transaction was active. Transaction should succeed if retried." ),
+        LockClientStopped( TransientError, "Lock client is stopped, no more locks can be acquired." ),
```
Too technical, I think. When and why are lock clients stopped? What is a user supposed to think when he gets this one popping out in a busy prod environment?

Commit:
Feature toggles now static final, rename termination reason method

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -312,7 +316,7 @@ public void markForTermination()
             {
                 failure = true;
-                terminated = true;
+                terminationReason = reason;
```
I'm suspecting the `terminationReleaseLock` should cover the entire body of this method (so this includes the `canBeTerminated` check), and the entire body of the `close()` method.

Calls to `release()` can otherwise sneak in between this `canBeTerminated` check, and our first load of the `reuseCount` field, allowing us to terminate a transaction that has been released back to the pool, and possibly lend back out again.

chrisvest replied:
If we do that, then I think all loads and stores with the `terminationReason` field will be protected by the lock, so the field won't need to be volatile.

burqen replied:
There is another problem where a tx that is being `markForTermination` while being `close` and the termination is not scheduled until the tx has returned to the pool and then initialised again, effectively terminating a completely different tx.

Here is a branch with test that reproduce this problem in a single thread by keeping the reference to the old `TopLevelTransaction` which holds the pooled `KernelTransactionImplementation`.
https://github.com/burqen/neo4j/tree/2.3-tx-terminate-after-reuse

I suggest grabbing the lock earlier in `markForTermination` and otherwise leave this problem to be adressed separate.

chrisvest replied:
Yes, let’s tackle this as a separate issue.

Cheers,
Chris

> On 01 Jul 2016, at 09:08, Anton Persson notifications@github.com wrote:
> 
> In community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java https://github.com/neo4j/neo4j/pull/7477#discussion_r69258346:
> 
> > @@ -312,7 +316,7 @@ public void markForTermination()
> >              if ( stillSameTransaction && canBeTerminated() )
> >              {
> >                  failure = true;
> > -                terminated = true;
> > -                terminationReason = reason;
> >   There is another problem where a tx that is being markForTermination while being close and the termination is not scheduled until the tx has returned to the pool and then initialised again, effectively terminating a completely different tx.
> 
> Here is a branch with test that reproduce this problem in a single thread by keeping the reference to the old TopLevelTransaction which holds the pooled KernelTransactionImplementation.
> https://github.com/burqen/neo4j/tree/2.3-tx-terminate-after-reuse https://github.com/burqen/neo4j/tree/2.3-tx-terminate-after-reuse
> I suggest grabbing the lock earlier in markForTermination and otherwise leave this problem to be adressed separate.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub https://github.com/neo4j/neo4j/pull/7477/files/e50e5ab9232bbc1050ccbfb3a4fa678df8f5aa8a#r69258346, or mute the thread https://github.com/notifications/unsubscribe/AAAfOcsNYBJInuCw5EvpQumFXQ7RxQhzks5qRLzRgaJpZM4JBmm9.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -774,7 +773,17 @@ private void release()
+        return terminationReason != null;
+    }
+
+    public long lastTransactionTimestampWhenStarted()
```
This guy is not part of the `KernelTransaction` interface? If it is, it needs an `@Override`

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java:
```diff
@@ -266,6 +281,39 @@ private void assertDatabaseIsRunning()
+    /**
+     * Allow new transactions to be started again if current thread is the one who called
+     * {@link #blockNewTransactions()}. If current thread is not the one that called {@link #blockNewTransactions()}
+     * or it has not been called at all then NO_OP.
```
Why do we allow unmatched block/unblock calls?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java:
```diff
@@ -266,6 +281,39 @@ private void assertDatabaseIsRunning()
     public KernelTransactionsSnapshot get()
     {
-        return new KernelTransactionsSnapshot( allTransactions );
+        return new KernelTransactionsSnapshot( allTransactions, Clock.SYSTEM_CLOCK.currentTimeMillis() );
```
We are using `Clocks` in a couple of places in here. Would it make sense to have it passed in?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -397,6 +401,10 @@ public Object relationshipGetProperty( long relationshipId, int propertyKeyId )
+        finally
+        {
+            statement.assertOpen();
+        }
```
Why do we have this check after we've read the property value? Property values aren't lazy. Same for `nodeGetProperty` above.

burqen replied:
Because if we read a dynamic property we can not trust the returned value to not be a mix of states caused my record reuse until we check if we are terminated or not.

```
try ( Transaction tx = db.beginTx() )
{
    String value = getSomeNode().getProperty( keyToLongString );
    // Value can not be trusted here
    tx.success();
}
```

This scenario is reproduced by `SlavesCanReadCorruptedData` (now renamed to `TerminationOfSlavesDuringPullUpdatesTest`)

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -439,6 +451,10 @@ public PrimitiveIntIterator relationshipGetPropertyKeys( long relationshipId ) t
+        finally
+        {
+            statement.assertOpen();
+        }
```
Why do we have this check at the end? Even if the iterator is lazy, this check will come too soon to protect anything, won't it?

burqen replied:
If the iterator is lazy, then yes you are correct in that it wont protect against anything.
If it is not we have the same scenario as above. Right?

chrisvest replied:
Yes. I think most of our iterators are lazy, so this is worth investigating.
The value is usually realised in the `hasNext` call of our implementations, which makes sense considering a `hasNext` that returns `true` is supposed to guarantee that `next` will return something. Thus you should be able to create wrappers that `assertOpen` as part of the `hasNext` calls.

Cheers,
Chris

> On 01 Jul 2016, at 09:48, Anton Persson notifications@github.com wrote:
> 
> In community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java https://github.com/neo4j/neo4j/pull/7477#discussion_r69261723:
> 
> > @@ -439,6 +451,10 @@ public PrimitiveIntIterator relationshipGetPropertyKeys( long relationshipId ) t
> >          {
> >              return relationship.get().getPropertyKeys();
> >          }
> > -        finally
> > -        {
> > -            statement.assertOpen();
> > -        }
> >   If the iterator is lazy, then yes you are correct in that it wont protect against anything.
> >   If it is not we have the same scenario as above. Right?
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub https://github.com/neo4j/neo4j/pull/7477/files/e50e5ab9232bbc1050ccbfb3a4fa678df8f5aa8a#r69261723, or mute the thread https://github.com/notifications/unsubscribe/AAAfOWz6EHIpCc4yHReGQLAjg6KZ7awjks5qRMZWgaJpZM4JBmm9.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionIdStore.java:
```diff
@@ -43,16 +43,47 @@
  */
 public interface TransactionIdStore
 {
+    // TODO: Document this somewhere.
```
You can complete it and add it as an implementation note in the javadoc for this interface.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionIdStore.java:
```diff
@@ -43,16 +43,47 @@
     long BASE_TX_ID = 1;
     long BASE_TX_CHECKSUM = 0;
+    long BASE_TX_COMMIT_TIMESTAMP = 0; // TODO: JAVADOCSA THIS
+    long UNKNOWN_TX_COMMIT_TIMESTAMP = 1; // TODO: JAVADOCSA THIS
```
Yes, please javadoc this.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionIdStore.java:
```diff
@@ -62,11 +93,12 @@
      * seen given to this method will be visible in {@link #getLastCommittedTransactionId()}.
      * @param transactionId the applied transaction id.
      * @param checksum checksum of the transaction.
+     * @param commitTimestamp
```
Explain the `commitTimestamp` parameter.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionIdStore.java:
```diff
@@ -96,13 +128,14 @@
+     *  @param transactionId transaction id that will be the last closed/committed id.
      * @param checksum checksum of the transaction.
-     * @param logVersion version of log the committed entry has been written into.
+     * @param commitTimestamp
```
Also here.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/RecoveryVisitor.java:
```diff
@@ -74,6 +75,7 @@ public boolean visit( RecoverableTransaction transaction ) throws IOException
         }
 
         lastTransactionIdApplied = txId;
+        lastTransactionCommitTimestamp = transaction.representation().getCommitEntry().getTimeWritten();
```
Since the grace period is only a heuristic, I suppose it doesn't matter if concurrent transactions are out of order in the log.

burqen replied:
Right, that should be the case as the potential commit time difference between the out of order transactions in the log should be very small in comparison to the `safeZone`.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/KernelTransactionImplementationTest.java:
```diff
@@ -458,11 +461,11 @@ public void markForTerminationTerminatedTransaction()
-        transaction.markForTermination();
+        transaction.markForTermination( Status.General.UnknownFailure );
+        transaction.markForTermination( Status.General.UnknownFailure );
+        transaction.markForTermination( Status.General.UnknownFailure );
```
You could additionally verify that the first reason isn't overwritten by later reasons.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/KernelTransactionsTest.java:
```diff
@@ -111,7 +122,7 @@ public void shouldDisposeTransactionsWhenAsked() throws Exception
         assertThat( postDispose, not( equalTo( second ) ) );
 
-        assertTrue( leftOpen.shouldBeTerminated() );
+        assertTrue( leftOpen.terminationReason() != null );
```
Don't we want to specify a particular reason?

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/legacylogs/LegacyLogsTest.java:
```diff
@@ -176,8 +188,65 @@ public void shouldRenameFiles() throws IOException
 
+    @Test
+    @SuppressWarnings( "unchecked" )
+    public void getTransactionInformation() throws IOException
```
This is not a very informative name for a test.

Commit:
Fixed test expecting wrong Exception

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -191,7 +201,86 @@ public void unpackResponse( Response<?> response, final TxHandler txHandler ) th
+         *              | |
+         *              | Start applying chunk
+         *              New tx starts here. Does not get terminated because not among active transactions, but will read
+         *              outdated data and can be affected by reuse contamination.
```
I'd like to see the two different parts of problem two explained separately. The diagram only explains `tx started after activeTransactions() but before apply`.

chrisvest replied:
Please also explain where in the code `activeTransactions()` is called, and how it relates to what goes on here.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -191,7 +201,86 @@ public void unpackResponse( Response<?> response, final TxHandler txHandler ) th
+        if ( batchSizeExceedsSafeZone() )
+        {
+            // Problem 2
+            kernelTransactions.blockNewTransactions();
```
Please explain what this has to do with Problem 2.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -191,7 +201,86 @@ public void unpackResponse( Response<?> response, final TxHandler txHandler ) th
+        if ( transactionQueue.isEmpty() )
+        {
+            return;
+        }
```
Why would the queue be empty?

lutovich replied:
Queue can be drained in two cases:
 1) it got full because batch of txs from master was too big
 2) processed response `#hasTransactionsToBeApplied()` returns `true`; this seem to always be true for `TransactionStreamResponse` regardless of transaction stream size

I think (2) is the reason for this check.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -191,7 +201,86 @@ public void unpackResponse( Response<?> response, final TxHandler txHandler ) th
+        finally
+        {
+            kernelTransactions.unblockNewTransactions();
+        }
```
Something like:

```
if ( batchSizeExceedsSafeZone() )
{
  kernelTransactions.blockNewTransactions();
  try
  {
    terminateUnsafeTransactionsAndApplyQueued();
  }
  finally
  {
    kernelTransactions.unblockNewTransactions();
  }
}
else
{
  terminateUnsafeTransactionsAndApplyQueued();
}
```

Would mean that you won't have unbalanced block/unblock methods, which is easier to follow.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -191,7 +201,86 @@ public void unpackResponse( Response<?> response, final TxHandler txHandler ) th
+        for ( KernelTransaction tx : kernelTransactions.activeTransactions() )
+        {
+            KernelTransactionImplementation kti = (KernelTransactionImplementation) tx;
+            long commitTimestamp = kti.lastTransactionTimestampWhenStarted();
```
Why not add that method to the `KernelTransaction` interface so we don't have to cast here?

chrisvest replied:
Also, `commitTimestamp` and `lastTransactionTimestampWhenStarted` sounds like two different things.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -239,13 +330,12 @@ private void applyQueuedTransactions() throws IOException
                         // at last closed transaction id they might get notified to continue before all data
                         // has actually been flushed properly.
-                        transactionQueue.accept( batchCloser );
+                        transactionQueue.accept( batchCloser ); // MARK TXs AS CLOSED
```
Be aware that this part is _completely_ different in 3.0. So there will likely be a lot of changes in to this and other classes in the merge commit.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -328,9 +420,99 @@ public void obligation( long txId ) throws IOException
+            //      - Flush buffered ids to disk (if memory intrusive) -- ??
+            //      - Only do assertOpen after property reads if we are slave or wait for transactions to terminate
+            //      - Place all comments describing the problem / solution in a better place
+            //      v Clean up how we handle Transient...Exception, TxTerminatedException, etc. + how we use Status.*
```
What about these TODO items?

chrisvest commented:
@lutovich @burqen General note: when you fix my comments, please do so in _new_ commits. Otherwise it'll be hard for me to keep track of what needs to be looked at, and what not.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -328,9 +420,99 @@ public void obligation( long txId ) throws IOException
+            // lifetime.
+            //
+            // PROTOTYPED SOLUTION
+            // Master and Slave agree on quarantine snapshotTime.
```
Except each one seems to just read their own configurations and just hope that everyone else in the cluster is configured with a similar "safe zone" duration.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -328,9 +420,99 @@ public void obligation( long txId ) throws IOException
+            //
+            // * Tx start on slave when slave is in SLAVE STATE 1
+            //      - Latest applied transaction has timestamp A and safe zone is A-T.
+            //      - Tx.startTime = A
```
So "startTime" is the time this transaction was committed? Because `A` is at the end of the transaction.
Also, given this information; what does it mean? Can this transaction commit, or not?

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -328,9 +420,99 @@ public void obligation( long txId ) throws IOException
+            //
+            // Scenario 1 - Tx finish when slave is in SLAVE STATE 2
+            //      Latest applied transaction in store has timestamp B and safe zone is B-T.
+            //      Tx did not cross the safe zone boundary as Tx.startTime = A > B-T
```
What? There is a literal line going from `A` through this transaction. How come it didn't cross any boundary?

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -328,9 +420,99 @@ public void obligation( long txId ) throws IOException
+            //      Latest applied transaction in store has timestamp C and safe zone is C-T.
+            //      We are just about to apply the next part of the batch and push the safe zone window forward.
+            //      This will make Tx.startTime = A < C-T. This means Tx is now in risk of reading mixed state records.
+            //      We will terminate Tx and let the user try again.
```
This is also confusing. Not sure the diagram I'm seeing is matching the text.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -328,9 +420,99 @@ public void obligation( long txId ) throws IOException
+            // |--------|----------|-->| ---|------>|----------->|  SLAVE
+            //
+            // tx started in S1 can finish in S2 but not in S3
+            // tx started in S2 can finish in S3.
```
This did not clarify anything for me.

chrisvest commented on enterprise/com/src/test/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpackerTest.java:
```diff
@@ -156,16 +160,55 @@ public void panicAndSkipBatchOnApplyingFailure() throws Throwable
+        InOrder inOrder = inOrder( kernelTransactions, applier );
+        inOrder.verify( applier, times( 1 ) ).closeBatch();
+        inOrder.verify( kernelTransactions, times( 1 ) ).unblockNewTransactions();
+    }
```
Do we also have a test for the observable behaviour of delayed freeing? Mock-based testing quite often relies a lot on implementation details, and the implementation of `TransactionCommittingResponseUnpacker` is completely different in 3.0, even though it exhibits pretty much the same behaviours. An implementation agnostic test would be less fragile, and make it easier to forward merge.

lutovich replied:
All tests in `TransactionCommittingResponseUnpackerTest` are also mock based. We have integration tests for this in `TerminationOfSlavesDuringPullUpdatesTest`.

chrisvest commented on enterprise/com/src/test/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpackerTest.java:
```diff
@@ -156,16 +160,55 @@ public void panicAndSkipBatchOnApplyingFailure() throws Throwable
+         * Tests that shutting down the response unpacker while in the middle of committing a transaction will
+         * allow that transaction stream to complete committing. It also verifies that any subsequent transactions
+         * won't begin the commit process at all.
+         */
```
This indentation needs fixing. Good comment, though.

chrisvest commented on enterprise/com/src/test/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpackerTest.java:
```diff
@@ -329,9 +382,16 @@ public void shouldThrowInCaseOfFailureToApply() throws Throwable
-                buildDependencies( logFile, logRotation, mock( IndexUpdatesValidator.class ), applier,
-                        appender, obligationFulfiller, kernelHealth ) );
+                deps );
+//        final TransactionCommittingResponseUnpacker unpacker = new TransactionCommittingResponseUnpacker( deps );
```
Doesn't look like you need this line.

chrisvest commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/factory/HighlyAvailableEditionModule.java:
```diff
@@ -161,6 +165,9 @@
         extends EditionModule
 {
+    public static final Setting<Long> id_reuse_safe_zone_time =
+            setting( "dbms.id_reuse_safe_zone", Settings.DURATION, "1h" );
```
Is this really the right place for this setting? What about `HaSettings`? You can even make it `@Internal` if you like.

chrisvest replied:
Also, if you want it to be hidden, you probably want to prefix it with `unsupported.`.

chrisvest commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -328,9 +420,99 @@ public void obligation( long txId ) throws IOException
+            // slave transaction can see states from all of the transactions that touches the reused record during its
+            // lifetime.
+            //
+            // PROTOTYPED SOLUTION
```
Is this still a prototype?

chrisvest commented on enterprise/ha/src/test/java/org/neo4j/kernel/ha/cluster/SlavesCanReadCorruptedData.java:
```diff
@@ -0,0 +1,450 @@
+import static org.neo4j.kernel.ha.factory.HighlyAvailableEditionModule.id_reuse_safe_zone_time;
+
+@RunWith( Parameterized.class )
+public class SlavesCanReadCorruptedData
```
This class does not say `Test` at the end, so I don't think it'll be picked up by sure-fire.

chrisvest commented:
@lutovich @burqen First review round done. Would be nice if there were fewer mocks in the tests. Also, from previous:

> General note: when you fix my comments, please do so in _new_ commits. Otherwise it'll be hard for me to keep track of what needs to be looked at, and what not.

Commit:
Grab tx termination lock in beginning of markForTermination

Commit:
KernelTransactions now take a Clock to use

Commit:
Set correct committed transaction timestamp

it is now collected from logs during startup if not present in store.

Also moved some javadoc and added tests

Commit:
Better javadoc for explaining the safeZone algorithm

Commit:
Better termination message from LockClient

Commit:
Added lastTransactionCommitTimestampWhenStarted to interface

Commit:
Dont use unmatched calls to block/unblock

lutovich committed:
Made it explicit that property keys are retrieved eagerly

lutovich committed:
Javadoc and test naming improvements

lutovich committed:
Improve assertion in a KTI test

lutovich committed:
Move `dbms.id_reuse_safe_zone` to HaSettings

lutovich committed:
Fix test and improve javadoc

lutovich committed:
Revert "Grab tx termination lock in beginning of markForTermination"

This reverts commit f2f783ca85d8e48284f788130b59c8c785fe5adb.
Grabbing lock earlier does not really solve the 'wrong transaction
termination' issue. Reverted commit made code forget to release lock in
some cases.

lutovich commented:
@chrisvest I think all comments except the first one should be addressed now. We'll change the first commit message before merging to not do rebase now and not loose comments.

lutovich committed:
Remember commit time when reading tx logs

Previously code only memorized transaction start entry.
Also fixed assertion based on objects `#toString()`.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionMetadataCache.java:
```diff
@@ -124,37 +135,18 @@ public boolean equals( Object o )
-            result = 31 * result + startPosition.hashCode();
-            result = 31 * result + (int) (checksum ^ (checksum >>> 32));
-            return result;
+            return Objects.hash( masterId, authorId, startPosition, checksum, timeWritten );
```
Looks like `equals` and `hashCode` are not strictly necessary on `TransactionMetadata` because they are never used as keys in hashed collections, or otherwise used in comparisons. Also, this mechanism of computing an objects hash code causes allocations – boxing for primitive elements, and then the carrier array itself – so I'd discourage it.

chrisvest commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/factory/HighlyAvailableEditionModule.java:
```diff
@@ -790,6 +795,41 @@ protected void doAfterRecoveryAndStartup( String editionName, DependencyResolver
+            }
+            catch ( IOException e )
+            {
+                throw new RuntimeException( "Unable to read transaction logs", e );
```
This should at least be `IllegalStateException`. Maybe there's another exception type that's even better.

Commit:
Better hashCode for TransactionMetadataCache

Commit:
More informative Exception from HighAvailabilityEditionModule

chrisvest commented:
@lutovich @burqen This looks good to merge once it's green; apparently one of the benchmarks depend on internal APIs, and fails to compile with these changes. I'm leaning toward changing the benchmark code to fix that, but I'll let you guys decide.

burqen commented:
@chrisvest Yes, changing the benchmark is what we plan to do.
